c/tskit/tables.c

```diff
@@ -5388,6 +5381,638 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+        fprintf(out, "L:%f, R:%f, P:%d, C:%d\n", self->tables->edges.left[j],
+            self->tables->edges.right[j], self->tables->edges.parent[j],
+            self->tables->edges.child[j]);
+    }
+    fprintf(out, "===\nNode table\n==\n");
+    for (j = 0; j < self->tables->nodes.num_rows; j++) {
+        fprintf(out, "ID:%f, Time:%f, Flag:%d\n", (double) j,
+            self->tables->nodes.time[j], self->tables->nodes.flags[j]);
+    }
+    fprintf(out, "==\nOldest parents\n==\n");
+    for (j = 0; j < self->tables->nodes.num_rows; j++) {
+        printf("For node %d: %d\n", (int) j, self->oldest_parent[j]);
+    }
+    fprintf(out, "==\nSample pairs\n==\n");
+    for (j = 0; j < 2 * self->num_pairs; j++) {
+        printf("%i ", (int) self->pairs[j]);
+        if (j % 2 != 0) {
+            printf("\n");
+        }
+    }
+    fprintf(out, "===\nAncestral map\n==\n");
+    for (j = 0; j < self->tables->nodes.num_rows; j++) {
+        fprintf(out, "Node %d: ", (int) j);
+        for (u = self->ancestor_map_head[j]; u != NULL; u = u->next) {
+            fprintf(out, "(%f,%f->%d)", u->left, u->right, u->node);
+        }
+        fprintf(out, "\n");
+    }
+    fprintf(out, "===\nIBD segments\n===\n");
+    for (j = 0; j < self->num_pairs; j++) {
+        fprintf(out, "Pair (%i, %i)\n", (int) self->pairs[2 * j],
+            (int) self->pairs[2 * j + 1]);
+        for (u = self->ibd_segments_head[j]; u != NULL; u = u->next) {
+            fprintf(out, "(%f,%f->%d)", u->left, u->right, u->node);
+        }
+        fprintf(out, "\n");
+    }
+}
+
+int TSK_WARN_UNUSED
+tsk_ibd_finder_run(tsk_ibd_finder_t *self)
+{
+
+    const tsk_edge_table_t *input_edges = &self->tables->edges;
+    int ret = 0;
+    size_t j;
+    tsk_id_t current_parent, u;
+    size_t num_edges = input_edges->num_rows;
+    tsk_segment_t *seg;
+    tsk_segment_t *s;
+    double intvl_l, intvl_r, current_time;
+    bool parent_should_be_added = true;
+
+    if (num_edges > 0) {
+        for (j = 0; j < num_edges; j++) {
+
+            // Stop if the processed node's time exceeds the max time.
+            current_parent = input_edges->parent[j];
+            current_time = self->tables->nodes.time[current_parent];
+            if (current_time > self->max_time) {
+                goto out;
+            }
+
+            // Extract segment.
+            seg = ibd_finder_alloc_segment(self, input_edges->left[j],
+                input_edges->right[j], input_edges->child[j]);
+            // Create a SegmentList holding all of the sample segments descending from
+            // seg.
+            u = seg->node;
+            if (self->is_sample[u]) {
+                ret = ibd_finder_enqueue_segment(self, seg->left, seg->right, seg->node);
+                if (ret != 0) {
+                    goto out;
+                }
+            } else {
+                if (self->ancestor_map_head[u] != NULL) {
+                    s = self->ancestor_map_head[u];
```

Comment:
for loop is a bit more concise:
```c
for (s = self->ancestor_map_head[u]; s != NULL; s = s->next) {
}
```
￼Also looks like you don't need the ``if (self->ancestor_map_head[u] != NULL) {`` condition either.